README.md
# Authorization

Manages everything related to the Authorization process of our Creditcard Processing Platform.

## Architecture

![Overal Architecture](./images/authorization.drawio.png)

## Documentation
The API is documented in openAPI 3.0 specs. These specs are kept up-to-date manually and are served by the application. This is integrated in the pipeline. So if you change something in the API DON'T FORGET to update the specs as well.

## Run the application local
- Checkout and run (the Local Development repo)(https://gitlab.cmpayments.local/creditcard/infrastructure/local-development).
- Thats it. :)

### Integration tests
Integration tests are ran using [Insomnia](https://insomnia.rest/download), kept in sync using [Repo Sync Plugin v3](https://insomnia.rest/plugins/insomnia-plugin-repo-sync-3).

Use the "Import from File" feature to import `tests/Authorization API.json`; this will import the suite and setup the sync plugin.

To push your changes, use `Repo Sync - Export Workspace`. This will write your current workspace into `tests/Authorization API.json`. Now use `git` to share the changes with the team, the same as you do with code changes.

To load changes, use `Repo Sync - Import Workspace`.

In addition to creating the required API calls in the `debug` panel, setup "unit tests" in the `test` panel, including checking specific response fields.

#### VTS Setup
To version the VTS testcases, a different integration suite is created per section of the integration test. These are stored in `tests/visa/vts-vip/*.stf`

### Application Logic
All Application logic resides inside `./cmd/api/`. These files can be built into a binary which services a REST API that
can be used to Authorize Creditcard payments.

### Domain Logic
All Domain logic resides in `./internal/processing/`. This directory has a package for every functional vertical:

* [`authorize`](./internal/processing/authorization) manages the process of authorizing
* [`refund`](./internal/processing/refund) manages the process of authorizing a refund
* [`capture`](./internal/processing/capture) manages the process of proceeding to actually get the money
* [`reverse`](./internal/processing/reverse) manages the process of reversing a payment authorization

Each of these directories contains all the directly required code. It does not contain concrete implementations
that use external dependencies (eg database queries, API clients, etc), these are hidden behind `interfaces`.

Details of the functionality of each of these verticals is described in a dedicated readme file inside each directory.

### Supporting / Infrastructure Code
All code that isn't directly needed for the domain, but is needed to support the processes around the domain,
has its own directory inside `./internal/`. Some prominent examples include:

* `config` to manage PSP and Merchant configuration
* `mip` and `eas` to connect to the Mastercard and VISA network
* [`spanner`](./internal/infrastructure/spanner/README.md)

### Set human readable logging for development
Edit your run configuration and add `human_readable_logging=true` to the environment variables in config.yml.

### Connection pool
To support parallel requests and responses a connection pool is used to connect to the different payment networks. More
details about this component can be found here [connection pool](./internal/infrastructure/connection/readme.md)

### Setting up CORS
CORS setup can be found in cmd/api/middlewares.go. For local development we use http://frontend.dev.cmtest.nl:3000/. When developing in a CM-environment any *.dev.cmtest.nl path will resolve to localhost by default. This means that technically anything.dev.cmtest.nl would work, but we have decided to use frontend.dev.cmtest.nl. All CM APIs have this CORS-rule enabled by default, so you cannot access any of the CM APIs if you try to approach it via localhost:3000.

For acceptance and production, this value should be set to whatever the acceptance/production backoffice urls are. The values can be found in the config.yml.

CHANGELOG.md
Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - yyyy-mm-dd
 
Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.
 
### Added

- [CA-1125](https://cmcom.atlassian.net/browse/CA-1125)
  Added changes for Magstripe transactions inside the domain

### Changed

### Fixed

- [CA-1154](https://cmcom.atlassian.net/browse/CA-1154)
  Fixed check if CAVV for Visa is a numeric value
- [CA-1156](https://cmcom.atlassian.net/browse/CA-1156)
    ECI 910 invalid for Mastercard


## [v0.0.61] - 2023-07-01
 
### Added

- Added test cards for acceptance tests


## [v0.0.60] - 2023-06-16
 
### Added

- [CA-975](https://cmcom.atlassian.net/browse/CA-975)
  Adding the tests from postman collection to authorization
- [CA-1144](https://cmcom.atlassian.net/browse/CA-1144)
  Download and parse Visa ARDEF table

### Fixed

- [CA-1105](https://cmcom.atlassian.net/browse/CA-1105)
  Fix for returning correct error when decoding CAVV

## [v0.0.59] - 2023-05-15

Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.

### Fixed

- [CA-1120](https://cmcom.atlassian.net/browse/CA-1120)
  Comment out Visa BIN parsing. This blocks starting the authorization repo.

## [v0.0.58] - 2023-05-15

Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.

### Added

- [CA-1080](https://cmcom.atlassian.net/browse/CA-1080)
  Added visa bin recognition

- [CA-833](https://cmcom.atlassian.net/browse/CA-833)
  Add echo endpoints to Swagger docs

- [CA-1081](https://cmcom.atlassian.net/browse/CA-1081)
  Added bin-blocking

- [CA-1106](https://cmcom.atlassian.net/browse/CA-1106)
  Added validation currency check

- [CA-1075](https://cmcom.atlassian.net/browse/CA-1075)
  Added check for acquiring EEA only

### Changed
- [CA-1100](https://cmcom.atlassian.net/browse/CA-1100)
  Removed validation for authorizationType, made it optional for refunds and authorizations.

- [CA-1092](https://cmcom.atlassian.net/browse/CA-1092)
  Removed topic authorization.presentment.cleared.v1 and dangling dependencies

- [CA-244](https://cmcom.atlassian.net/browse/CA-244)
  Moved currency and country to platform and imported new version of platform.

### Fixed

- [CA-1093](https://cmcom.atlassian.net/browse/CA-1093)
  Reversal always Full, even if Partial Capture exists

## [v0.0.57] - 2023-05-02

### Added

- [CA-1067](https://cmcom.atlassian.net/browse/CA-1067)
  Mask 8-digit BINs.

### Changed
- [CA-862](https://cmcom.atlassian.net/browse/CA-862)
  Added new version of platform which contains the single place of mastercard category codes.
  Changed all the dependencies to point to the MCCS on platform.

### Fixed

 - [CA-1089](https://cmcom.atlassian.net/browse/CA-1089)
   Add panic recovery to PubSub subscribers


## [v0.0.56] - 2023-03-30

Fix critical bug in starting the authorization service.

### Fixed
- [CA-1074](https://cmcom.atlassian.net/browse/CA-1074)
  Fix path to BIN ranges and limit card file look back loop

## v0.0.55 - 2023-03-29

Expose echo endpoints for the NOC to monitor and fix showstopper bug in refunds.

### Added
 - [CA-833](https://cmcom.atlassian.net/browse/CA-833)
    Send echo messages to VISA EAS and Mastercard MIP from HA-Proxy and monitoring endpoint (`GET /v1/echo/<scheme>`)
 - [CA-832](https://cmcom.atlassian.net/browse/CA-832)
    Gather throughput metrics and expose them via `GET /v1/metrics`.

### Changed
 - [CA-226](https://cmcom.atlassian.net/browse/CA-226)
    Properly log Tokenization errors; do not return error details to client
 - [CA-190](https://cmcom.atlassian.net/browse/CA-190)
    Only accept API keys that are at least 32 bits
 - [CA-212](https://cmcom.atlassian.net/browse/CA-212)
    Move generic code from Authorization source repository into a shared library

### Fixed
 - [CA-267](https://cmcom.atlassian.net/browse/CA-267)
    Downgrade database migrations
 - [CA-1071](https://cmcom.atlassian.net/browse/CA-1071)
    Publish POS data for clearing


## [v0.0.54] - 2023-03-15

Last pre-1.0 version; baseline for the changelog.



cmd/visa/main.go
package main

import (
	"encoding/hex"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
	"gitlab.cmpayments.local/creditcard/authorization/internal/data"
	"gitlab.cmpayments.local/creditcard/authorization/internal/entity"
	"gitlab.cmpayments.local/creditcard/authorization/internal/processing/scheme/visa"
	"gitlab.cmpayments.local/creditcard/authorization/pkg/iso8583"
	"gitlab.cmpayments.local/creditcard/authorization/pkg/visa/base1"
	"gitlab.cmpayments.local/creditcard/platform/currencycode"
)

type config struct {
	host string
	port int
}

var (
	loglevel                         string
	version                          = "0.0.1"
	cfg                              config
	signon, echo, authorize, signoff bool
	sourceId                         string
	logError, logVerbose, logDebug   Logger = NopLogger, NopLogger, NopLogger
)

func main() {
	// the default address is the HA proxy address.
	// address of the actual EAS and its ports is documented on https://cmcom.atlassian.net/wiki/spaces/CA/pages/951287809/Visa+authorization
	//flag.StringVar(&cfg.host, "host", "10.18.27.52", "EAS server host")
	flag.StringVar(&cfg.host, "host", "mip.test.cmpayments.local", "EAS server host")
	flag.IntVar(&cfg.port, "port", 10106, "EAS server port")
	flag.StringVar(&loglevel, "log", "vvv", "log level:\n\tv = only errors are displayed,\n\tvv = opening and closing connection information text is displayed,\n\tvvv = response result is displayed")
	flag.BoolVar(&signon, "signon", false, "If a Signon should be sent")
	flag.BoolVar(&echo, "echo", true, "If an Echo should be sent")
	flag.BoolVar(&authorize, "authorize", false, "If a Authorization should be sent")
	flag.BoolVar(&signoff, "signoff", false, "If a Signoff should be sent")
	flag.StringVar(&sourceId, "sourceid", "111106", "VISA Source Station ID")

	flag.Parse()

	switch true {
	case strings.Contains("v", loglevel):
		logError = PrintLogger
	case strings.Contains("vv", loglevel):
		logError = PrintLogger
		logVerbose = PrintLogger
	case strings.Contains("vvv", loglevel):
		logError = PrintLogger
		logVerbose = PrintLogger
		logDebug = PrintLogger
	}

	log.Printf("binary version: %s", version)

	//resolve tcp address
	logVerbose("resolve tcp address %s:%d\n", cfg.host, cfg.port)
	addr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf("%s:%d", cfg.host, cfg.port))
	if err != nil {
		logError("failed to resolve TCP address: %s", err.Error())
		os.Exit(1)
	}

	//establish connection
	logVerbose("establish %s connection to %s:%d\n", "tcp", cfg.host, cfg.port)
	c, err := net.DialTCP("tcp", nil, addr)
	if err != nil {
		logError("error dialing server: %s", err.Error())
		os.Exit(1)
	}
	logVerbose("connection established")
	defer c.Close()

	//set connection properties
	err = c.SetKeepAlive(true)
	if err != nil {
		logError("failed to set keepalive: %s", err.Error())
		os.Exit(1)
	}
	err = c.SetKeepAlivePeriod(30 * time.Second)
	if err != nil {
		logError("failed to set keepalive period: %s", err.Error())
		os.Exit(1)
	}
	err = c.SetDeadline(time.Now().Add(15 * time.Second))
	if err != nil {
		logError("failed to set read/write deadline: %s", err.Error())
		os.Exit(1)
	}

	//time.Sleep(10 * time.Second)

	if signon {
		//construct sign-on message
		logVerbose("construct sign-on message")
		signOnMsg := &visa.Message{
			SourceStationID: sourceId,
			Mti:             iso8583.NewMti("0800"),
			Fields: base1.Fields{
				F007_TransmissionDateTime:             base1.F007FromTime(time.Now()),
				F011_SystemTraceAuditNumber:           "100001",
				F070_NetworkManagementInformationCode: "071",
			},
		}

		signOnReq, err := visa.NewRequest(signOnMsg).Packet()
		if err != nil {
			logError("error constructing message: %s", err.Error())
			os.Exit(1)
		}

		logVerbose("sign-on message: %x", signOnReq)

		//send sign-on message
		logVerbose("send sign-on message")
		_, err = c.Write(signOnReq)
		if err != nil {
			logError("error writing message: %s", err.Error())
			os.Exit(1)
		}

		//receive a response
		logVerbose("waiting for sign-on response")
		buffer := make([]byte, 1024)
		mLen, err := c.Read(buffer)
		if err != nil {
			logError("error reading: %s", err.Error())
			os.Exit(1)
		}

		received := visa.NewResponse(buffer, nil)
		visaMsg := received.(visa.Response)

		logDebug("sign-on response: %+v", visaMsg.Message())
		logDebug("sign-on response received: %s", hex.EncodeToString(buffer[:mLen]))
	}

	if echo {
		//construct echo message
		logVerbose("construct echo message")
		echo := visa.Echo(sourceId)

		echoReq, err := visa.NewRequest(echo).Packet()
		if err != nil {
			logError("error constructing message: %s", err.Error())
			os.Exit(1)
		}

		logVerbose("echo message: %x", echoReq)

		//send echo message
		logVerbose("send echo message")
		_, err = c.Write(echoReq)
		if err != nil {
			logError("error writing message: %s", err.Error())
			os.Exit(1)
		}

		//receive an echo response
		logVerbose("waiting for response")
		echoBuffer := make([]byte, 1024)
		mLen, err := c.Read(echoBuffer)
		if err != nil {
			logError("error reading: %s", err.Error())
			os.Exit(1)
		}

		received := visa.NewResponse(echoBuffer, nil)
		visaMsg := received.(visa.Response)

		logDebug("echo response: %+v", visaMsg.Message())
		logDebug("echo response received: %s", hex.EncodeToString(echoBuffer[:mLen]))
	}

	if authorize {
		hours, minutes, _ := time.Now().Clock()
		stanStr := fmt.Sprintf("00%02d%02d", hours, minutes)
		stan, _ := strconv.Atoi(stanStr)
		a := &entity.Authorization{
			ID:                       uuid.UUID{},
			LogID:                    uuid.UUID{},
			Amount:                   123,
			Currency:                 currencycode.Must(currencycode.EUR),
			CustomerReference:        "1234",
			Source:                   "moto",
			LocalTransactionDateTime: data.LocalTransactionDateTime(time.Now()),
			Status:                   "",
			Stan:                     stan,
			InstitutionID:            fmt.Sprintf("%06d", 20814),
			ProcessingDate:           time.Now(),
			CreatedAt:                time.Now(),
			Card: entity.Card{
				Number: "4761340000000035",
				Holder: "Test",
				Expiry: entity.Expiry{
					Year:  "23",
					Month: "12",
				},
			},
			CardAcceptor: entity.CardAcceptor{
				CategoryCode: "0742",
				ID:           "1234",
				Name:         "VTS Test",
				Address: entity.CardAcceptorAddress{
					PostalCode:  "4899AL",
					City:        "Breda",
					CountryCode: "NLD",
				},
			},
		}

		visa.AuthorizationSchemeData(a)

		authMsg, err := visa.MessageFromAuthorization(*a)
		if err != nil {
			logError("error defining visa message: %w", err.Error())
			os.Exit(1)
		}
		authMsg.SourceStationID = sourceId
		req := visa.NewRequest(authMsg)

		packet, err := req.Packet()
		if err != nil {
			logError("error creating auth packet: %s", err.Error())
			os.Exit(1)
		}
		_, err = c.Write(packet)
		if err != nil {
			logError("error writing auth message: %s", err.Error())
			os.Exit(1)
		}

		//receive a response
		logVerbose("waiting for auth response")
		authBuff := make([]byte, 1024)
		mLen, err := c.Read(authBuff)
		if err != nil {
			logError("error auth reading: %s", err.Error())
			os.Exit(1)
		}

		logVerbose("auth response received: %s", hex.EncodeToString(authBuff[:mLen]))

		received := visa.NewResponse(authBuff, nil)
		visaMsg := received.(visa.Response)
		logDebug("visa message F062.SF1: %#v", visaMsg.Message().Fields.F062_CustomPaymentServiceFields.SF1_AuthorizationCharacteristicsIndicator)
		logDebug("visa message F062.SF2: %#v", visaMsg.Message().Fields.F062_CustomPaymentServiceFields.SF2_TransactionIdentifier)
		if visaMsg.Error() != nil {
			logError("visa error: %s", visaMsg.Error().Error())
		}
	}

	if signoff {
		//construct sign-off message
		logVerbose("construct sign-off message")
		signOffMsg := &visa.Message{
			SourceStationID: sourceId,
			Mti:             iso8583.NewMti("0800"),
			Fields: base1.Fields{
				F007_TransmissionDateTime:             base1.F007FromTime(time.Now()),
				F011_SystemTraceAuditNumber:           "100001",
				F070_NetworkManagementInformationCode: "072",
			},
		}

		signOffReq, err := visa.NewRequest(signOffMsg).Packet()
		if err != nil {
			logError("error constructing message: %s", err.Error())
			os.Exit(1)
		}

		//send sign-off message
		logVerbose("send sign-off message")
		_, err = c.Write(signOffReq)
		if err != nil {
			logError("error writing message: %s", err.Error())
			os.Exit(1)
		}

		//receive a response
		logVerbose("waiting for response")
		signOffBuf := make([]byte, 1024)
		mLen, err := c.Read(signOffBuf)
		if err != nil {
			logError("error reading: %s", err.Error())
			os.Exit(1)
		}

		received := visa.NewResponse(signOffBuf, nil)
		visaMsg := received.(visa.Response)

		logVerbose("sign-off response: %+v", visaMsg.Message())
		logVerbose("sign-off response received: %s", hex.EncodeToString(signOffBuf[:mLen]))
	}

	os.Exit(0)
}

type Logger func(format string, a ...any)

func PrintLogger(format string, a ...any) {
	log.Printf(format, a...)
}

func NopLogger(_ string, _ ...any) {}
